sources/common/inc/SafeUInt.h

```diff
@@ -0,0 +1,221 @@
+/// in all copies or substantial portions of the Software.
+///
+/// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+/// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+/// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+/// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+/// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+/// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+/// IN THE SOFTWARE.
+
+#pragma once
+
+/// @defgroup Generic Generic
+///
+/// @brief Generic classes.
+///
+/// @par Full Description
+/// The group contains classes which could be used anywhere else.
+
+#include <cstdint>
+#include <limits>
+#include <iostream>
+#include <stdexcept>
+
+/// A safe unsigned integer class.
+/// The overridden arthmetic operators throw exceptions on the overflow.
+template<class T> class SafeUInt
+{
+    public:
+        /// The safe integer constructor.
+        ///
+        /// @param value The value of the integer.
+        SafeUInt(T value) :
+            m_Value(value)
+        {
+        }
+
+        /// The default constructor.
+        ///
+        /// By default the integer is set to 0.
+        SafeUInt() :
+            m_Value(0)
+        {
+        }
+
+        /// Returns the integer value.
+        T GetValue() const
+        {
+            return m_Value;
+        }
+
+        /// Overrides 'equal to' operator.
+        bool operator==(const SafeUInt<T> &rhs) const
+        {
+            return m_Value == rhs.GetValue();
+        }
+
+        /// Overrides 'not equal to' operator.
+        bool operator!=(const SafeUInt<T> &rhs) const
+        {
+            return !(*this == rhs);
+        }
+
+        /// Overrides 'greater than' operator.
+        bool operator>(const SafeUInt<T> &rhs) const
+        {
+            return (m_Value > rhs.GetValue());
+        }
+
+        /// Overrides 'less than' operator.
+        bool operator<(const SafeUInt<T> &rhs) const
+        {
+            return (m_Value < rhs.GetValue());
+        }
+
+        /// Overrides the addition operator '+'.
+        SafeUInt<T> operator+(const SafeUInt<T> &rhs) const
+        {
+            if ( (std::numeric_limits<T>::max() - m_Value) < rhs.GetValue())
+            {
+                throw std::out_of_range("");
+            }
+            return SafeUInt<T>(m_Value + rhs.GetValue());
+        }
+
+        /// Overrides the addition operator '+='.
+        SafeUInt<T> &operator+=(const SafeUInt<T> &rhs)
+        {
+            if ((std::numeric_limits<T>::max() - m_Value) < rhs.GetValue())
+            {
+                throw std::out_of_range("");
+            }
+            m_Value += rhs.GetValue();
+            return *this;
+        }
+
+        /// Overrides the substraction operator '-'.
+        SafeUInt<T> operator-(const SafeUInt<T> &rhs) const
```

Comment:
Very bad. `-` operator overload should be expressed using `-=` and should not be a member.